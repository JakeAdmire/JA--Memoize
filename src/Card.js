import React, { Component } from 'react';
import Answers from './Answers.js';

export default class Card extends Component {
  constructor(props) {
    super(props);
    this.state = { showAnswer: false, attemptCorrect: false };
  }
  showResults = (event) => {
    if (event.target.innerText === this.props.frontContent.solutionPrototype) {
      this.setState({attemptCorrect: true, showAnswer: true});
    } else { 
      this.setState({attemptCorrect: false, showAnswer: true})
    }
  }
  changeCards = (event) => {
    if (event.target.classList.value === 'left') { 
      this.props.prevCard() 
    } else { 
      this.props.nextCard() 
    }
    this.setState({showAnswer: false})
  }
  handleClick = (event) => {
    console.log(event.target.innerText);
    if (event.target.innerText == 'try again') { this.setState({showAnswer: false}); 
    } else { 
      this.props.nextCard();
      this.setState({showAnswer: false, attemptCorrect: false})
    }
  }
  generateCard() {
    let header = this.state.attemptCorrect ? 'Good Job!' : 'Incorrect!';
    let button = this.state.attemptCorrect ? 'next prototype' : 'try again';
    if (this.state.showAnswer) {
      return (
      <div className="card-back">
        <h2>{header}</h2>
        <button onClick={this.handleClick}>{button}</button> 
      </div>
      ) 
    } else {
      return (
      <div className="card-front">
        <h3>What prototype would you use to..</h3>
        <h3>{this.props.frontContent.problem}</h3> 
        <h4 className="snippet">{this.props.frontContent.snippet}</h4>
      </div>
      )
    }
  }
  render() {
    let cardClass = this.state.showAnswer ? 'back' : '';
    return (
      <div className={`card ${cardClass}`}>
        <button onClick={this.changeCards} className="left">PREVIOUS
          <span></span>
        </button>
        <button onClick={this.changeCards} className="right">SKIP
          <span></span>
        </button>
        <span onClick={() => this.setState({showAnswer: !this.state.showAnswer})} className='card-counter'>{this.props.cardNumber + 1}</span>
        {this.generateCard()}
        <Answers  answers={this.props.answers}
                  correctAnswer={this.props.correctAnswer}
                  showResults={this.showResults}/>
      </div>
    )
  }
}