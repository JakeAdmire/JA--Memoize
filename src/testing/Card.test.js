import React from 'react';
import ReactDOM from 'react-dom';
import Card from '../Card';
import { shallow } from 'enzyme';

const mockCard = {problem: ''};
const showPrevCardMock = jest.fn();
const showNextCardMock = jest.fn();

describe('Card', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(
      <Card 
          frontContent={mockCard}
          nextCard={showNextCardMock}
          prevCard={showPrevCardMock}
        />
    );
  });
  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should have default state', () => {
    expect(wrapper.state()).toEqual({ 
      showAnswer: false,
      attemptCorrect: false,
      cardNumber: 0
     });
  })

  it('should invoke handleClick on button click', () => {
    wrapper.instance().setState({'showAnswer': true});
    wrapper.find('.back-button').simulate('click', {target: {innerText: 'try again'}});
    expect(wrapper.state('showAnswer')).toEqual(false);
  })

  it('should invoke changeCards on left button click', () => {
    wrapper.instance().setState({'showAnswer': true});
    expect(wrapper.state('showAnswer')).toEqual(true);
    wrapper.find('.left').simulate('click', {target: {classList: {value: 'left'}}});
    expect(wrapper.state('showAnswer')).toEqual(false);
  })

  it('should invoke changeCards on right button click', () => {
    wrapper.instance().setState({'showAnswer': true});
    expect(wrapper.state('showAnswer')).toEqual(true);
    wrapper.find('.right').simulate('click', {target: {classList: {value: 'right'}}});
    expect(wrapper.state('showAnswer')).toEqual(false);
  })

  it('should change state when showResults is invoked', () => {
    expect(wrapper.state('showAnswer')).toEqual(false);
    wrapper.instance().showResults({target: {innerText: 'sort'}});
    expect(wrapper.state('showAnswer')).toEqual(true);
  })

})