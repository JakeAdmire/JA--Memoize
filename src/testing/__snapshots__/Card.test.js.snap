// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card should match snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Card
    frontContent={
      Object {
        "problem": "",
      }
    }
    nextCard={[MockFunction]}
    prevCard={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <button
          className="left"
          onClick={[Function]}
        >
          PREVIOUS
          <span />
        </button>,
        <button
          className="right"
          onClick={[Function]}
        >
          SKIP
          <span />
        </button>,
        <span
          className="card-counter"
        >
          1
        </span>,
        <div
          className="card-front"
        >
          <h3>
            What prototype would you use to..
          </h3>
          <h3>
            
          </h3>
          <h4
            className="snippet"
          />
        </div>,
        <Answers
          showResults={[Function]}
        />,
      ],
      "className": "card ",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "PREVIOUS",
            <span />,
          ],
          "className": "left",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": Array [
          "PREVIOUS",
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": "span",
          },
        ],
        "type": "button",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "SKIP",
            <span />,
          ],
          "className": "right",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": Array [
          "SKIP",
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": "span",
          },
        ],
        "type": "button",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": 1,
          "className": "card-counter",
        },
        "ref": null,
        "rendered": 1,
        "type": "span",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <h3>
              What prototype would you use to..
            </h3>,
            <h3>
              
            </h3>,
            <h4
              className="snippet"
            />,
          ],
          "className": "card-front",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "What prototype would you use to..",
            },
            "ref": null,
            "rendered": "What prototype would you use to..",
            "type": "h3",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "",
            },
            "ref": null,
            "rendered": "",
            "type": "h3",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": undefined,
              "className": "snippet",
            },
            "ref": null,
            "rendered": null,
            "type": "h4",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "answers": undefined,
          "correctAnswer": undefined,
          "showResults": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <button
            className="left"
            onClick={[Function]}
          >
            PREVIOUS
            <span />
          </button>,
          <button
            className="right"
            onClick={[Function]}
          >
            SKIP
            <span />
          </button>,
          <span
            className="card-counter"
          >
            1
          </span>,
          <div
            className="card-front"
          >
            <h3>
              What prototype would you use to..
            </h3>
            <h3>
              
            </h3>
            <h4
              className="snippet"
            />
          </div>,
          <Answers
            showResults={[Function]}
          />,
        ],
        "className": "card ",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "PREVIOUS",
              <span />,
            ],
            "className": "left",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": Array [
            "PREVIOUS",
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": "span",
            },
          ],
          "type": "button",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "SKIP",
              <span />,
            ],
            "className": "right",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": Array [
            "SKIP",
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": "span",
            },
          ],
          "type": "button",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": 1,
            "className": "card-counter",
          },
          "ref": null,
          "rendered": 1,
          "type": "span",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <h3>
                What prototype would you use to..
              </h3>,
              <h3>
                
              </h3>,
              <h4
                className="snippet"
              />,
            ],
            "className": "card-front",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "What prototype would you use to..",
              },
              "ref": null,
              "rendered": "What prototype would you use to..",
              "type": "h3",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "",
              },
              "ref": null,
              "rendered": "",
              "type": "h3",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": undefined,
                "className": "snippet",
              },
              "ref": null,
              "rendered": null,
              "type": "h4",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "answers": undefined,
            "correctAnswer": undefined,
            "showResults": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
